EIPS/eip-7412.md

```diff
@@ -0,0 +1,167 @@
+---
+eip: 7412
+title: On-Demand Off-Chain Data Retrieval
+description: A method to construct multicalls with prepended verifiable off-chain data
+author: Noah Litvin (@noahlitvin), db (@dbeal-eth)
+discussions-to: https://ethereum-magicians.org/t/erc-7412-on-demand-off-chain-data-retrieval/15346
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-07-26
+---
+
+## Abstract
+
+Contracts may require off-chain data that is not already available on chain during execution. A smart contract function could implement the standard proposed here by reverting with `error OracleDataRequired(address oracleContract, bytes oracleQuery)`. Clients supporting this standard would recognize this error message during a simulation of the request, query the specified decentralized oracle network for signed data, and instead stage a transaction with a multicall that prepends the verification of the required off-chain data. The data would be written on-chain during verification to a smart contract for the subsequent call to read, avoiding the error.
+
+## Motivation
+
+Ethereum's scaling roadmap involves a series of separate execution contexts for smart contract code (including layer two and layer three scaling solutions). This makes the ability to read data across multiple chains crucial to the construction of scalable applications. Also, for decentralized finance protocols that rely on price data, it is not reasonable to expect oracle networks will be able to continuously push fresh data to every layer two and layer three network for an arbitrary number of price feeds.
+
+Cross-chain bridges are being developed, adhering to standards such as [ERC-6170](./eip-6170.md), where smart contract functions can write data to other chains. There is a need for a similar standard that enables reading data from other chains. This standard can be generalized for reading any off-chain data from a decentralized oracle network, including price feeds.
```

Comment:
You are allowed to reference other proposals, **but** doing so means your proposal cannot advance past their statuses. So, for example, if ERC-6170 is a Draft, this proposal wouldn't be able to advance to Review.

If you're fine with that, you can leave this reference in. I just wanted to point it out.